include/parlay/type_traits.h

```diff
@@ -230,56 +246,183 @@ inline constexpr bool is_trivial_allocator_v = is_trivial_allocator<Alloc, T>::v
-                           (std::is_nothrow_move_constructible<T>::value &&
-                            std::is_nothrow_destructible<T>::value)> { };
+template<typename T>
+using is_trivially_relocatable = std::is_trivially_relocatable<T>;
+
+// Clang's builtin with some (very limited) support for trivial relocatability
+#elif defined(__has_builtin)
+#if __has_builtin(__is_trivially_relocatable)
+
+template<typename T>
+struct is_trivially_relocatable : std::bool_constant<__is_trivially_relocatable(T)> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#else
+
+template<typename T>
+struct is_trivially_relocatable : std::is_trivially_copyable<T> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#endif
+#else
+
+template<typename T>
+struct is_trivially_relocatable : std::is_trivially_copyable<T> { };
+
+#define PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE
+
+#endif
 
 template<typename T>
 inline constexpr bool is_trivially_relocatable_v = is_trivially_relocatable<T>::value;
 
+#if defined(__cpp_impl_trivially_relocatable)
+#define PARLAY_TRIVIALLY_RELOCATABLE [[trivially_relocatable]]
+#else
+#define PARLAY_TRIVIALLY_RELOCATABLE
+#endif
+
+
+#if defined(PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE)
+
+#define PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(...)                              \
+  struct is_trivially_relocatable<__VA_ARGS__> : std::true_type {};
+
+#define PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF(cond, ...)                     \
+  struct is_trivially_relocatable<__VA_ARGS__> : std::bool_constant<cond> {};
+
+
+// We list all the types that ought to be trivially relocatable under any
+// sensible implementation. It is possible that some of these are wrong.
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::allocator<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::shared_ptr<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::weak_ptr<T>);
+
+template<typename T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE(std::optional<T>);
+
+template<typename T, typename Deleter>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<Deleter>), std::unique_ptr<T, Deleter>);
+
+template<typename T1, typename T2>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<T1> &&
+                                        is_trivially_relocatable_v<T2>), std::pair<T1, T2>);
+
+template<typename T, std::size_t N>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<T>), std::array<T, N>);
+
+template <typename... T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<T> && ...), std::tuple<T...>);
+
+template <typename... T>
+PARLAY_ASSUME_TRIVIALLY_RELOCATABLE_IF((is_trivially_relocatable_v<T> && ...), std::variant<T...>);
+
+
+// Container implementations can differ widely by vendor, so we don't want to specialize these
+// with a broad brush.  Instead, only specialize them for stdlibs that we are confident about.
```

Comment:
FYI, https://quuxplusone.github.io/blog/2019/02/20/p1144-what-types-are-relocatable/ — although I'd encourage you to do _less_ of this warranting-of-types-you-don't-own, because this is exactly how Folly got in trouble. 😛 Of course it makes a difference for performance (_iff_ you're not using a P1144-enabled STL implementation to begin with). The good news is that you've guarded this all under `PARLAY_MUST_SPECIALIZE_IS_TRIVIALLY_RELOCATABLE`, so if there are any bugs, my P1144 implemention will _fix_ them rather than cause them. ;)